### ISSUE TITLE
Backup files in batches to avoid timeouts

### ISSUE THREAD

**Issue by pdewouters** [humanmade/backupwordpress#752]
Right now when we start a backup, the process can generally take more than 30 seconds which is usually the PHP timeout limit.
We could use the file manifest method to start a backup and time it. Each batch should not run for more than 10 - 20 seconds to be safe. We keep note of where we are so we also can resume backup if it fails

We would create a subfolder per backup with the archive parts in it. Instead of manually creating the multiple zip files, let's use zip's split size functionality.
` zip -s 670m -r foo bar`

Doesn't seem that you can do delta backups at the same time though. I do think that this method would be more beneficial if we can avoid timeouts.

>  Use  -s  to set the split size and create a split archive.  The size is
       given as a number followed optionally by one of k (kB), m (MB), g (GB),
       or  t (TB) (the default is m).  The -sp option can be used to pause zip
       between splits to allow changing removable media, for example, but read
       the descriptions and warnings for both -s and -sp below.

       Though  zip does not update split archives, zip provides the new option
       -O (--output-file or --out) to allow split archives to be  updated  and
       saved in a new archive.  For example,

              zip inarchive.zip foo.c bar.c --out outarchive.zip

**Comment by willmot** [comment]
If we have access to `zip` then we won't be hitting PHP timeouts as time taken performing system commands isn't counted towards the PHP execution time.

Only `PclZip` and `ZipArchive` should be hitting PHP timeouts.

Even if we are hitting PHP timeouts with `zip` it would be caused by something other than the actual `zip` process and so splitting into delta wouldn't help.

**Comment by pdewouters** [comment]
ah, doesn't running `exec` or `shell_exec` pause execution until it gets a response?

**Comment by pdewouters** [comment]
Just did a little digging and yes, you're right.It doesn't hit a PHP timeout error, but it can hit the server gateway timeout error.
Also, memory usage can be intensive

**Comment by willmot** [comment]
Using `zip`s `-s` flag to create a split archive wouldn't help with either the time it takes nor the memory usage as it still does the whole thing in a single process, it's just that it creates multiple files instead of a single file.

To do what you are suggesting we'd need to batch the backup process at the PHP process level and spawn each smaller backup job as a separate Backdrop task. That way each would be in it's own thread and would use it's own memory etc. There's no need to do this for `zip` as it doesn't cause PHP timeouts / memory issues, like you say it can cause proxy timeouts, but those are are harmless, just because the proxy has timed out doesn't mean the zip process has stopped. But it could be worth it for ZipArchive and Pclzip, that way we could avoid PHP timeouts / memory limit error on servers with low limits.

This is what WP Remote did with it's file manifest approach, it first created a text file containing all the files on the server and then ran through it in batches of 100 and added them to the zip, restarting itself each time.

**Comment by pdewouters** [comment]
Instead of a file limit, I'd prefer to use a timer, so we know we stay under the timeout limit

**Comment by willmot** [comment]
Yeah agree a timer makes sense, we could even try to detect their timeout limit and dynamically set the timeout to something less than that.

**Comment by pdewouters** [comment]
This would also fix issues with PHP timing out on uploading large zip files to remote servers

**Comment by owaincuvelier** [comment]
My thoughts on this are; Is this the primary solution to fix chunked uploads to third party locations? Will the development time pay off? Because Daniel tried this before with limited success, and also I'm wary of introducing more issues then this could potentially solve. 
Do we want to solve backups at a comprehensive level as VaultPress say?

**Comment by pdewouters** [comment]
@owaincuvelier valid concerns. my understanding was that on WP Remote there was still only one zip file per backup, is that not the case?

Individual file syncing like VaultPress does would require significant development, basically a rewrite of the plugin.

**Comment by pdewouters** [comment]
What if user has uploaded a large file > 200MB - this would potentially still cause timeout when transferring.

**Comment by owaincuvelier** [comment]
>my understanding was that on WP Remote there was still only one zip file per backup, is that not the case?

There was indeed only one Zip per backup. I would be interested to see how a multi archive would work functionally. But from a users perspective this gets potentially intimidating. 

>Individual file syncing like VaultPress does would require significant development, basically a rewrite of the plugin.

I wouldn't suggest we do this, but this is a comprehensive solution to backups, what I'm suggesting is we just accept the fact BackUpWordPress isn't going to work for everyone, so we make it work great for those it does work. Not to say that this goes agains't that, but if this doesn't solve chunked uploads, does this become a rabbit hole?

**Comment by pdewouters** [comment]
Actually the ZipArchive process does this already. It does it in batches of 500 files.
As mentioned already, the zip process doesn't have the issue.

**Comment by willmot** [comment]
It adds files to the backup in batches of 500 but it does that all in the same thread. Backing up in batches is about splitting to multiple threads so that each one is shorter than the timeout limit.

**Comment by pdewouters** [comment]
ah yes. we could use a similar approach to this https://pippinsplugins.com/batch-processing-for-big-data/
or use the same approach as we do for calculating the folder/filesizes?

**Comment by pdewouters** [comment]
It seems that if we nail this, a lot of open issues will be resolved.
I think this feature and https://github.com/humanmade/backupwordpress/issues/525 could be priorities

**Comment by willmot** [comment]
The issue with this approach is that it relies on the ability to create new PHP threads which means using Backdrop and relying even more on wp-cron.

Either we do this and further increase our dependency on a working wp-cron. Or we don't and instead work to reduce our dependency (which is what we're discussing in https://github.com/humanmade/backupwordpress/issues/821). I'm tempted most by the former, IE we require a working wp-cron to function. This will continue to be a large support burden though as it's so easy to break wp-cron (some hosts don't support loopback requests, maintenance mode plugins, `.htaccess` etc).

**Comment by pdewouters** [comment]
ah yes, good point - I was only thinking about manual backups

**Comment by pdewouters** [comment]
Creating multipart backups could be a separate feature.
